### PR TITLE
refactor(compartment-mapper): use @endo/errors [mirror of endojs/endo#3152]

### PR DESCRIPTION
**Pull request by kriscendobot**
Mirror of upstream PR `endojs/endo#3152` by @erights, opened in `endojs/endo-but-for-bots` for offline review and CI iteration.

Refs: endojs/endo#3152, endojs/endo#3110

## Description

This PR moves from the `assert` style towards the `expr || Fail\`...\`` style in `packages/compartment-mapper/src/compartment-map.js`, which is more efficient in the typical non-erroneous case and is more idiomatic for Endo. It introduces a new dependency from `@endo/compartment-mapper` to `@endo/errors`. If it is important for `@endo/compartment-mapper` not to transitively depend on `ses`, then this PR should probably wait on endojs/endo#3110 or a successor that accomplishes the same goal.

The bot-mirror's `master` lags upstream `endojs/endo` master by ~132 commits at the time of this mirror. The upstream branch was rooted at upstream commit `83c0b06fe4` (a March 30 2026 commit predating the bot fork's most recent sync). The PR's substantive surface is three files: `packages/compartment-mapper/package.json` (one new dependency), `packages/compartment-mapper/src/compartment-map.js` (the refactor), and `yarn.lock`. The aggregate file diff against `bots-but-for-bots`'s `master` will include 132 commits of unrelated drift; reviewers should look at the upstream PR's diff or filter to the three files above.

### Security Considerations

None.

### Scaling Considerations

If this PR waits for endojs/endo#3110 or equivalent, then none. Otherwise, the transitive dependency on `ses` may be significant. The existing use of `ses`'s `assert` shows that `@endo/compartment-mapper` is already transitively dependent, so probably makes no difference either way.

### Documentation Considerations

We want to encourage the `... || Fail...` style over the `assert` style, so good to get rid of violations of our own style preferences in our own code.

### Testing Considerations

All existing tests pass without modification, which should be good enough for this refactoring PR.

### Compatibility Considerations

Some error messages that were formed with untagged template literals are now formed with `Fail`. Untagged template literals don't redact, which seems to be intentional here. To preserve that intention, this PR annotates all `Fail` interpolations that did not already have a `q(...)` with a `b(...)`.

### Upgrade Considerations

None.

## Iteration plan

A juror panel and saboteur run here; review feedback is addressed via follow-up commits on top of the upstream branch tip and cherry-picked back to upstream when the maintainer accepts. The upstream branch's tip (`a8f1243f3b`) is preserved as-is at the base of this mirror so cherry-picks back to upstream remain trivial.

### CODE REVIEW

**Review by kriscendobot — COMMENTED** [review]
## Panel review (8 perspectives, mirror of endojs/endo#3152)

cc @kriskowal. Submitting via `--comment` because the bot is the gh-side author of this mirror PR; treat this as a `request-changes` for triage purposes. Findings here are intended to be relayed to upstream `endojs/endo#3152` (or addressed as follow-up commits on top of `a8f1243f3b`).

### Reviewer 1: Correctness

**Verdict:** comment-only

**Findings:**

- The PR converts six `assert(...)` / `assert.typeof(...)` call sites in `compartment-map.js` to `expr || Fail\`...\`` style, but leaves ten un-converted (`assert(...)` calls at lines 356, 463, 616, 620, 781, 791 of the post-PR file). These un-converted sites all interpolate the imported `q(...)` from `@endo/errors` into a plain string template (e.g., line 358, 619). The interpolation works because `q(value)` returns an object with a meaningful `toString()` (`bestEffortStringify(value)`), but the rendered output differs from the previous `JSON.stringify(value)`: `bestEffortStringify` knows how to render `bigint`, `undefined`, symbols, and remotables, while `JSON.stringify` returns `undefined` (which then interpolates as the string `"undefined"`) or throws (`bigint`). For the un-converted sites, the values being quoted are URL strings, keypath strings, and string arrays, so the format change is invisible in practice. Worth noting in case the maintainer wants follow-up consistency.
- `assertConditions` (line 147 to 155 post-PR) chose a different conversion shape than its siblings: `if (!isArray(conditions)) { throw Fail\`...\`; }` rather than `isArray(conditions) || Fail\`...\``. Both are correct (the `throw` is redundant because `Fail` itself throws, but TypeScript's CFA may need the `throw` to narrow). The five sibling conversions in this PR all use the `expr || Fail\`...\`` pattern; assertConditions is the lone outlier. Pick one shape per file.

**Notes (out of scope):**

- None.

### Reviewer 2: Type-system enforcement

**Verdict:** request-changes

**Findings:**

- The PR base predates upstream commit `ec42cb7b8f` (`chore(types): typescript 'strict' conformance`) which wrapped the four `errors.push(error)` calls in lines 321 to 346 with `/** @type {Error} */ (error)` casts. The PR branch lacks those casts. When this PR rebases onto current `actual/master`, the conflict will mostly auto-merge, but the resulting file will need the strict-mode casts re-applied or the build will fail under `tsconfig`'s `strict` flag (now enabled on master). This is a peer-fix-missed concern.
- `@returns {asserts allegedLabel is string}` on line 92 corrects a pre-existing typo (was `alleged is string` referring to a non-existent parameter). Good catch but bundled into a refactor commit; would have been a cleaner two-commit history.

**Notes (out of scope):**

- The `assertTruthy` overload (lines 462 to 469) has a latent bug unrelated to this PR: when `url` is undefined, the assertion message is set to `url` itself (which is undefined). Pre-existing.

### Reviewer 3: Error-message quality

**Verdict:** comment-only

**Findings:**

- `assertEmptyObject` (line 138 to 140 post-PR) wraps the entire pre-built `message` argument in `b(message)`. Callers pass strings like `\`${keypath} must not have extra properties; got ${q(extra)} in ${q(url)}\`` (built outside this function). The interpolated `q(extra)` is already a `MarshalQuoteWrapper` whose `toString()` produces the rendered form; the resulting string then becomes the bare-interpolation. Effectively, all of the message (including the interpolated values that the caller intended to be safely-quoted) is now treated as a single bare-tagged value. The redaction story this PR claims ("annotates all `Fail` interpolations that did not already have a `q(...)` with a `b(...)`") is technically met but in spirit the bare-wrap-of-pre-formatted-string defeats the per-value redaction discipline. Consider passing the parts through and tagging at the leaf rather than wrapping a pre-built message.
- `assertConditions` line 150 uses `${b(String(conditions))}` to render the offending value. `String(conditions)` throws on a Symbol (`Cannot convert a Symbol value to a string`); the original was `${conditions}` (implicit string coercion, which also throws on Symbol). So no regression, but a switch to `${q(conditions)}` (which uses `bestEffortStringify` and handles Symbol gracefully) would be more robust.

**Notes (out of scope):**

- None.

### Reviewer 4: Naming and prose (style guide compliance)

**Verdict:** comment-only

**Findings:**

- `stringCompare = (a, b) => ...` now shadows the imported `b` (the bare quoter), and the conversion adds `// eslint-disable-next-line no-shadow` to silence the rule. The shadow is real and the disable is the cheapest fix, but a reader at the call site sees `b` and has to context-switch to know which `b` is in play. Cleaner alternatives: rename the parameter (`(left, right)`, or `(x, y)`), or rename the import (`import { b as bare } from '@endo/errors'`). The rest of the file uses `b(...)` heavily, so renaming the parameter is the lower-friction fix.
- The comment on lines 4 to 7 of the pre-PR file (`// TODO convert to the new \`||\` assert style. Deferred because this file pervasively uses simple template strings ...`) was removed. The TODO is now partially satisfied; consider replacing with `// TODO convert remaining \`assert(...)\` sites in this file to \`||\` style for full consistency` to mark the remaining work.

**Notes (out of scope):**

- None.

### Reviewer 5: Diff hygiene

**Verdict:** comment-only

**Findings:**

- The PR mixes three concerns: (a) the substantive `assert` → `Fail` refactor, (b) a typo fix in a JSDoc `@returns asserts` clause (line 92), (c) addition of `// eslint-disable-next-line no-shadow` to silence a rule the import introduced. All three could have been separate commits; bundling them is minor but harms reviewability. The PR is small enough that the cost is low.
- `yarn.lock` change is bundled with the source commit. The lockfile is conventionally its own `chore: Update yarn.lock` commit so it can be dropped and regenerated cleanly on rebase. The upstream PR has only one commit; splitting would need the maintainer's cooperation.

**Notes (out of scope):**

- None.

### Reviewer 6: Test coverage

**Verdict:** comment-only

**Findings:**

- No tests in `packages/compartment-mapper/test/` directly exercise the `compartment-map.js` validators (`assertString`, `assertLabel`, etc.). They are exercised transitively through bundle / archive tests. The PR claims "All existing tests pass without modification", which is true: `npx ava` in the package reports 803 passed, 6 known failures (pre-existing). No tests pin the specific format of the error messages this PR rewrites, so the format change from `JSON.stringify` to `bestEffortStringify` (in the un-converted `assert` sites that reuse the imported `q`) is invisible to the test suite. A regression-evidence experiment is therefore not directly possible at the unit level; this is a refactor with no behavioral assertion to verify.

**Notes (out of scope):**

- A dedicated unit test file for `compartment-map.js` validators would catch future regressions in the assertion style. Not in scope for this PR.

### Reviewer 7: Performance

**Verdict:** approve

**Findings:**

- The PR's stated motivation is performance: `expr || Fail\`...\`` short-circuits in the non-erroneous case, while `assert(expr, msg)` always evaluates `msg` (the template literal interpolation runs even when the assertion passes). The conversion is correct: `expr || Fail\`...\`` only invokes the tagged template when `expr` is falsy. The six converted call sites are all on hot validation paths (called per compartment / per module / per scope during compartment-map validation), so the win is real.
- The remaining un-converted `assert(...)` sites still pay the format-the-message cost on every call. Following up to convert those would extend the win.

**Notes (out of scope):**

- None.

### Reviewer 8: Regression evidence and rebase safety

**Verdict:** request-changes

**Findings:**

- The PR branch (`actual/markm-compartment-mapper-errors`, tip `a8f1243f3b`) is rooted at `83c0b06fe4` (March 30 2026). Two substantive commits to `compartment-map.js` have landed on `actual/master` since then:
  - `ac8b01ae98` (`fix(compartment-mapper): cull underscore-prefixed properties from serialized compartment maps`) modified `getModuleConfigurationSpecificProperties` to filter keys starting with `_` rather than destructuring three known keys. The PR branch retains the old destructuring shape. After rebase, the conflict resolution must keep the master version of this function.
  - `ec42cb7b8f` (`chore(types): typescript 'strict' conformance`) added `/** @type {Error} */ (error)` casts to four `errors.push(error)` calls. The PR branch lacks these casts. Without them, a strict-mode TypeScript build will error.
- The PR will not merge cleanly onto current upstream master without a rebase that re-applies these two fixes; the maintainer should be aware. This is a sibling-fork-misses-peer-fix concern.

**Notes (out of scope):**

- The bot mirror's `master` lags upstream by ~132 commits, so this finding cannot be reproduced locally on the mirror. Verify against `actual/master` directly.

## Aggregated verdict

**Must fix before upstream merge:**

1. Rebase onto current `actual/master` and re-apply the strict-mode `/** @type {Error} */ (error)` casts from `ec42cb7b8f` and the underscore-cull fix from `ac8b01ae98`. (Reviewer 2 + 8.)

**Should fix in this PR (or relayed to upstream):**

1. Rename the `stringCompare` parameter `b` to avoid shadowing the imported `b`, and remove the `no-shadow` eslint-disable. (Reviewer 4.)
2. Choose one shape between `expr || Fail` and `if (!expr) throw Fail` and use it consistently in `assertConditions`. (Reviewer 1.)
3. Reconsider `assertEmptyObject`'s wrap-the-whole-message-in-`b(...)` pattern; the bare-tag should annotate per-value, not the pre-formatted string. (Reviewer 3.)
4. Replace the removed `TODO convert to the new || assert style` comment with a residual TODO citing the still-unconverted sites, or convert them in this PR for full file-level consistency. (Reviewer 4.)

**Out of scope / follow-up:**

1. The latent `assertTruthy` bug (`url` becomes the message when undefined) is pre-existing.
2. A unit test file for `compartment-map.js` validators would catch format regressions; separate PR.

The mirror PR's tip is preserved at upstream `a8f1243f3b` so cherry-picks back to upstream remain trivial. Follow-up commits to address these findings would land on top of the mirror; relay to upstream as appropriate.

**Comment by kriskowal** [comment]
Please dispatch a fixer to rebase on master (synced to actual/master) address above feedback, including tests for edge cases identified by the saboteur, and then shepherd through CI.

File: packages/compartment-mapper/src/compartment-map.js [hunk 1]
```diff
@@ -32,6 +32,9 @@ import {
  *   DigestedCompartmentDescriptor} from './types.js'
  */
 
+// TODO convert remaining `assert(...)` sites in this file to `||` style
+// for full consistency.
+
```
Comment (kriskowal) [inline packages/compartment-mapper/src/compartment-map.js]: Please finish this job.

**Review by kriskowal — CHANGES_REQUESTED** [review]

File: packages/compartment-mapper/src/compartment-map.js (hunk 1 above)
Comment (kriscendobot) [reply packages/compartment-mapper/src/compartment-map.js]: Done. Converted the 6 remaining `assert(...)` sites in `compartment-map.js` to `expr || Fail`...`` (assertModuleConfiguration line 344, assertTruthy line 451, assertFileUrlString lines 600 & 604, assertCompartmentDescriptors lines 753 & 763) and dropped the residual TODO. Lint clean; all 876 compartment-mapper tests pass (6 known failures unchanged). New HEAD f99922441a1334e55360c9ddf906cc3825deae06.
